.github/workflows/nightly-t5x-build.yaml

```diff
@@ -22,40 +22,86 @@ permissions:
 jobs:
 
   metadata:
-    if: (github.event_name == 'workflow_run' && github.event.workflow_run.conclusion == 'success') || github.event_name == 'workflow_dispatch'
     runs-on: ubuntu-22.04
     outputs:
+      PUBLISH: ${{ steps.if-publish.outputs.PUBLISH }}
       BUILD_DATE: ${{ steps.date.outputs.BUILD_DATE }}
     steps:
+      - name: Check if the triggering workflow failed
+        id: if-upstream-failed
+        shell: bash -x -e {0}
+        run: |
+          echo "UPSTREAM_FAILED=${{ github.event.workflow_run.conclusion != 'success' }}" >> $GITHUB_OUTPUT
+
+      - name: Cancel workflow if upstream workflow did not success
+        if: ${{ steps.if-upstream-failed.outputs.UPSTREAM_FAILED == 'true' }}
+        uses: styfle/cancel-workflow-action@0.12.0
```

Comment:
I see there's new canceling logic. What does this do that the previous implementation didn't?

Reply:
Potentially allowing for more complicated cancellation logic in the metadata job, plus nicer job graph visualization IMHO :)

Reply:
Previously the logic was limited to what a single-line GitHub Actions expression allows. Now we can do arbitrary bash scripting to determine whether a job needs to be cancelled.

Reply:
Got it. That's helpful. IIRC, I think there were sometimes race conditions when `if-upstream-failed`'s outcome was "fail", since it wasn't listed in any job's `needs:` section, you could sometimes get farther along on jobs dependent on the `metadata` job before they fail.